From c611073cc8e694529520e5c3a27e5a8083591c48 Mon Sep 17 00:00:00 2001
From: yangxiaoxing <yangxiaoxing@xiaomi.com>
Date: Mon, 15 Feb 2016 20:18:49 +0800
Subject: [PATCH] BatteryStats: fix wrong calculation of battery voltage

there is a bug that the battery voltage of batterystats is twice as
the actual one, and it's because that the way to compose battery info
is different from the way to decompose.

Change it to the following format:

|-----7bit-----|-----10bit-----|-----14bit-----|-----1bit-----|
  battery level  battery temp    battery volt	 flag

Change-Id: I6181d77a7127d6e531d99898ea54ae9b62cbd39a
Signed-off-by: yangxiaoxing <yangxiaoxing@xiaomi.com>
---

diff --git a/core/java/com/android/internal/os/BatteryStatsImpl.java b/core/java/com/android/internal/os/BatteryStatsImpl.java
index 64b7768..2dece5f 100644
--- a/core/java/com/android/internal/os/BatteryStatsImpl.java
+++ b/core/java/com/android/internal/os/BatteryStatsImpl.java
@@ -1968,8 +1968,8 @@
 
     private int buildBatteryLevelInt(HistoryItem h) {
         return ((((int)h.batteryLevel)<<25)&0xfe000000)
-                | ((((int)h.batteryTemperature)<<14)&0x01ff8000)
-                | ((((int)h.batteryVoltage)<<1)&0x00007fff);
+                | ((((int)h.batteryTemperature)<<15)&0x01ff8000)
+                | ((((int)h.batteryVoltage)<<1)&0x00007ffe);
     }
 
     private int buildStateInt(HistoryItem h) {
@@ -2111,8 +2111,8 @@
         if ((firstToken&DELTA_BATTERY_LEVEL_FLAG) != 0) {
             batteryLevelInt = src.readInt();
             cur.batteryLevel = (byte)((batteryLevelInt>>25)&0x7f);
-            cur.batteryTemperature = (short)((batteryLevelInt<<7)>>21);
-            cur.batteryVoltage = (char)(batteryLevelInt&0x3fff);
+            cur.batteryTemperature = (short)((batteryLevelInt<<7)>>22);
+            cur.batteryVoltage = (char)((batteryLevelInt>>1)&0x3fff);
             cur.numReadInts += 1;
             if (DEBUG) Slog.i(TAG, "READ DELTA: batteryToken=0x"
                     + Integer.toHexString(batteryLevelInt)
